client/engine/engine.go

```diff
@@ -745,25 +751,25 @@ func fromMsgErr(id protocols.ObjectiveId, err error) error {
 }
 
 // getProposalObjectiveId returns the objectiveId for a proposal.
-func getProposalObjectiveId(p consensus_channel.Proposal) protocols.ObjectiveId {
+func getProposalObjectiveId(p consensus_channel.Proposal) (protocols.ObjectiveId, error) {
 	switch p.Type() {
 	case consensus_channel.AddProposal:
 		{
 			const prefix = virtualfund.ObjectivePrefix
 			channelId := p.ToAdd.Guarantee.Target().String()
-			return protocols.ObjectiveId(prefix + channelId)
+			return protocols.ObjectiveId(prefix + channelId), nil
 
 		}
 	case consensus_channel.RemoveProposal:
 		{
 			const prefix = virtualdefund.ObjectivePrefix
 			channelId := p.ToRemove.Target.String()
-			return protocols.ObjectiveId(prefix + channelId)
+			return protocols.ObjectiveId(prefix + channelId), nil
 
 		}
 	default:
 		{
-			panic("invalid proposal type")
```

Comment:
Here's another example when we would really like the type system to help us out. It may be possible to refactor this so there are no errors and no panics.

Reply:
Here's one idea in that direction https://github.com/statechannels/go-nitro/commit/1ec641f2bace4620020e167576d64bbbdaa10eb7

Reply:
The lack of exhaustive pattern matching in Go is a real shame though https://news.ycombinator.com/item?id=9037557

Reply:
> Here's one idea in that direction https://github.com/statechannels/go-nitro/commit/1ec641f2bace4620020e167576d64bbbdaa10eb7

Using a boolean would restrict us to two values, which works here, but I'm not sure we can do something similar if we have more than 2 types.  In fact instead of using a switch statement, we could just use if/else.

In general, I feel like this approach might be too clever for own good? I feel like we're trying to trick `golang` into having exhaustive pattern matching instead of working within the constraints of the language.

>This one is potentially debatable. If we hit this line, something has gone very badly wrong and it is definitely a programming error rather than something unexpected in the environment. The data in the store is in an invalid state. So one could argue that panicking is acceptable here.

I think you're right and we should just panic. When I first made these changes I wasn't really paying too much attention to the context of the `panic`. 

We don't ever expect a client to hit that panic and if they do something is seriously wrong and we should be panicking. I would advocate for continuing to use panics in exhaustive switch statements.

Reply:
Yup, will do!